File: .github/scripts/deadline-check-org.mjs
import { Octokit } from "@octokit/rest";
import fetch from "node-fetch";

const octokit = new Octokit({
  auth: process.env.PERSONAL_ACCESS_TOKEN, // Use the token passed from the workflow
  request: { fetch },
});

const org = "" //ORGANIZATION NAME HERE

const { data: repos } = await octokit.repos.listForOrg({
  org: org,
  type: "all",
});

const getDateDifferenceInDays = (date1, date2) => {
  const normalizedDate1 = new Date(date1.getFullYear(), date1.getMonth(), date1.getDate());
  const normalizedDate2 = new Date(date2.getFullYear(), date2.getMonth(), date2.getDate());

  // Calculate the difference in days
  const differenceInDays = (normalizedDate2 - normalizedDate1) / (1000 * 60 * 60 * 24);

  return differenceInDays;
};

for (const repo of repos) {
  console.log(`Checking repository: ${repo.name}`);
  await checkDeadlines(org, repo.name);
}



async function checkDeadlines(owner, repo) {
  try {
    const { data: issues } = await octokit.issues.listForRepo({
      owner,
      repo,
      state: "open",
    });

    const today = new Date();

    for (const issue of issues) {
      console.log(`Checking issue #${issue.number}: ${issue.title}`);

      const assignees = issue.assignees.map((assignee) => `@${assignee.login}`).join(", ");
      const deadlineLabel = issue.labels.find(label => label.name.startsWith("deadline:"));
      const reviewerLabel = issue.labels.find(label => label.name.startsWith("reviewer:"));

      if (deadlineLabel) {
        const deadlineString = deadlineLabel.name.replace("deadline:", "").trim();
        const deadlineDate = new Date(deadlineString.replace(/-/g, "/"));

        const daysLeft = getDateDifferenceInDays(today, deadlineDate);

        console.log(`Issue #${issue.number} has a deadline in ${daysLeft} days`);

        // Notify if the deadline is approaching or today
        if (daysLeft === 7 || daysLeft === 1 || daysLeft === 0) {
          let message;
          if (daysLeft === 0) {
            message = "⏰ Today is the deadline for this issue!";
          } else {
            message = `⏰ Reminder: This issue is due in ${daysLeft} day(s).`;
          }

          await octokit.issues.createComment({
            owner,
            repo,
            issue_number: issue.number,
            body: `${assignees} ${message}`,
          });
          console.log(`Comment posted on issue #${issue.number}`);
        }

        // Notify if the deadline has passed
        if (daysLeft < 0) {
          if (reviewerLabel) {
            const reviewer = reviewerLabel.name.replace("reviewer:", "").trim();
            const message = `@${reviewer} The deadline for this issue has passed.`;

            await octokit.issues.createComment({
              owner,
              repo,
              issue_number: issue.number,
              body: message,
            });
            console.log(`Late notification sent to reviewer ${reviewer} for issue #${issue.number}`);
          } else {
            console.log(`No reviewer label found for overdue issue #${issue.number}`);
          }
        }
      } else {
        console.log(`No deadline label found on issue #${issue.number}`);
      }
    }
  } catch (error) {
    console.error("Error checking deadlines:", error);
  }
}

File: .github/scripts/deadline-check-org.py
import requests
import os
from datetime import datetime, timedelta

# GitHub API Token
GITHUB_TOKEN = os.getenv("PERSONAL_ACCESS_TOKEN")
HEADERS = {"Authorization": f"token {GITHUB_TOKEN}"}

# Organization and repository details
ORG = ""  # Replace with organization name
OWNER = "ABenoit0226"  # Replace with owner name
REPO = "Deadline-Notification-Bot"  # Replace with repository name


def get_repos(org):
    url = f"https://api.github.com/orgs/{org}/repos"
    response = requests.get(url, headers=HEADERS)
    if response.status_code == 200:
        return response.json()
    else:
        print("Error fetching repositories:", response.json())
        return []


def get_issues(owner, repo):
    url = f"https://api.github.com/repos/{owner}/{repo}/issues?state=open"
    response = requests.get(url, headers=HEADERS)
    if response.status_code == 404:
        print(f"Repository {owner}/{repo} not found. Check if the repository name is correct.")
        return []
    try:
        issues = response.json()
        if isinstance(issues, list):
            return issues
        else:
            print("Unexpected response format for issues:", issues)
            return []
    except ValueError:
        print("Failed to parse JSON response for issues.")
        return []


def post_comment(owner, repo, issue_number, message):
    url = f"https://api.github.com/repos/{owner}/{repo}/issues/{issue_number}/comments"
    data = {"body": message}
    response = requests.post(url, json=data, headers=HEADERS)
    return response.json()


def get_date_difference_in_days(date1, date2):
    return (date2 - date1).days


def check_deadlines(owner, repo):
    today = datetime.today()
    issues = get_issues(owner, repo)
    
    for issue in issues:
        if not isinstance(issue, dict):
            print("Skipping unexpected issue format:", issue)
            continue
        
        issue_number = issue.get("number")
        title = issue.get("title", "Unknown Title")
        print(f"Checking issue #{issue_number}: {title}")
        
        assignees = ", ".join([f"@{assignee['login']}" for assignee in issue.get("assignees", [])])
        labels = {label["name"]: label["name"] for label in issue.get("labels", [])}
        
        deadline_label = next((label for label in labels if label.startswith("deadline:")), None)
        reviewer_label = next((label for label in labels if label.startswith("reviewer:")), None)
        
        if deadline_label:
            deadline_string = deadline_label.replace("deadline:", "").strip()
            try:
                deadline_date = datetime.strptime(deadline_string, "%Y-%m-%d")
            except ValueError:
                print(f"Invalid date format for issue #{issue_number}: {deadline_string}")
                continue
            
            days_left = get_date_difference_in_days(today, deadline_date)
            
            print(f"Issue #{issue_number} has a deadline in {days_left} days")
            
            if days_left in [7, 1, 0]:
                message = "⏰ Today is the deadline for this issue!" if days_left == 0 else f"⏰ Reminder: This issue is due in {days_left} day(s)."
                post_comment(owner, repo, issue_number, f"{assignees} {message}")
                print(f"Comment posted on issue #{issue_number}")
            
            if days_left < 0 and reviewer_label:
                reviewer = reviewer_label.replace("reviewer:", "").strip()
                post_comment(owner, repo, issue_number, f"@{reviewer} The deadline for this issue has passed.")
                print(f"Late notification sent to reviewer {reviewer} for issue #{issue_number}")
        else:
            print(f"No deadline label found on issue #{issue_number}")


if ORG:
    repos = get_repos(ORG)
    for repo in repos:
        if isinstance(repo, dict) and "name" in repo:
            print(f"Checking repository: {repo['name']}")
            check_deadlines(ORG, repo['name'])
        else:
            print("Skipping invalid repository format:", repo)
elif OWNER and REPO:
    check_deadlines(OWNER, REPO)
